services/brig/src/Brig/Data/Client.hs

```diff
@@ -199,7 +214,17 @@ claimPrekey u c = withOptLock u c $ do
               . field "client" (toByteString c)
               . msg (val "last resort prekey used")
       return $ Just (ClientPrekey c (Prekey i k))
-    Nothing -> return Nothing
+    removeAndReturnPreKey Nothing = return Nothing
+
+    pickRandomPrekey :: [(PrekeyId, Text)] -> AppIO (Maybe (PrekeyId, Text))
+    pickRandomPrekey [] = return Nothing
+    -- unless we only have one key left
+    pickRandomPrekey [pk] = return $ Just pk
+    -- pick among list of keys, except lastPrekeyId
+    pickRandomPrekey pks = do
+      let pks' = filter (\k -> fst k /= lastPrekeyId) pks
+      ind <- liftIO $ randomRIO (0, length pks' - 1)
```

Comment:
Ah, one more useless pedanticism: I think cryptonite puts more effort into their PRNGs, but this use is sensitive to attacks against randomness, so it should be fine.

Reply:
This is actually a good point. @jmatsushita what PRNG is used here? It's not much of a security issue here that's why I didn't think about it. But with a bad PRNG we of course increase the likelihood of errors seen by clients.

Reply:
It uses `randomRIO` from the `random` package https://hackage.haskell.org/package/random-1.2.0/docs/System-Random.html which uses the `splitmix` algorithm https://hackage.haskell.org/package/splitmix which is advertised as safe for randomized data structures (which I think is our case) but not for cryptographic uses. I would think it's a good fit for our use case, but just say the word and I'll change it to use Cryptonite's PRNG.

Reply:
I'm wondering now: what's a good definition of a non-cryptographic PRNG?

(No need to change it, I think.)

Reply:
Non-cryptographic RNGs like this one can for example be predictable. That's fine in a case like this, but fatal when used for any cryptographic operation. Most non-cryptographic RNGs trade this and other security properties for speed.
I don't see the need for a cryptographic one and the one used here appears to be good enough.
One attack that's for example possible with the predictability of this RNG is that an attacker can, after learning the sequence, try to produce a collision for a specific target. But this far from straight forward and would only lead to the victim having to re-establish the attacked session.
So I agree, no need to change it.

Reply:
Non-cryptographic RNGs like this one can for example be predictable. That's fine in a case like this, but fatal when used for any cryptographic operation. Most non-cryptographic RNGs trade this and other security properties for speed.
I don't see the need for a cryptographic one and the one used here appears to be good enough.
One attack that's for example possible with the predictability of this RNG is that an attacker can, after learning the sequence, try to produce a collision for a specific target. But this far from straight forward and would only lead to the victim having to re-establish the attacked session.
So I agree, no need to change it.

Reply:
Excellent, thanks a lot for the review :)